### PR TITLE
Update readme md

### PR DESCRIPTION
**Pull request by rustem-net**
Fix for issue #6 which adds a much needed introduction to this project.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -2,6 +2,118 @@
 
 Option is a C# representation of the maybe monad
 
+### Monad
```
Comment (rodoherty1) [inline README.md]: To be honest, I would consider removing this description of a Monad.  There are plenty of descriptions of a Monad all around the web.  In this case, I don't think it helps explain this library.

File: README.md [hunk 2]
```diff
@@ -2,6 +2,118 @@
...
+Monads are used to encapsulate a value and simplify handling of unusual cases. You can think of monads as generic types.
+
+### Maybe
+Maybe is a data type (also known as an option type) which is used to represent an optional value.
+
+### Maybe Monad
+Maybe monad is a generic type which contains either some value of it's underlying type or nothing.
+
+If you need more information you can check [wikibooks introduction](https://en.wikibooks.org/wiki/Haskell/Understanding_monads/Maybe) to maybe monad.
```
Comment (rodoherty1) [inline README.md]: Consider removing the "If you need" text and instead attach the hyperlink to the words "Maybe monad".

File: README.md [hunk 3]
```diff
@@ -2,6 +2,118 @@
...
+# Why we need the Option library?
+
+> For simplicity you can think of `Nullable<T>` being closest thing to the maybe monad in C#. Although it doesn't represent all the characteristics of the maybe monad, it's sufficient enough to give the main idea.
+
+At first glance, extending `Nullable<T>` might be a good option, 
+but the `struct` constraint for `T` in it makes it impossible to use with the reference types.
+
+Using maybe monad also makes code more readable by reducing the nested code blocks, as you will see in the following examples.
+
+# How to use it
```
Comment (rodoherty1) [inline README.md]: I would consider moving this paragraph up to the top of the README.
I believe the first thing people want to see is the code because that answers so many questions.

* What imports do I need?
* Is the syntax nice?
* Do I need a maths degree to use this library?
* Will it make my code more or less maintainable?

File: README.md [hunk 4]
```diff
@@ -2,6 +2,118 @@
...
+
+Our hardworking teammates are trying to publish the Option library to [nuget](https://www.nuget.org/), but until then you can download/clone/fork this repository and build it locally.
```
Comment (rodoherty1) [inline README.md]: Perhaps rephrase this so that it references the ticket that has been opened against this.

File: README.md [hunk 5]
```diff
@@ -2,6 +2,118 @@
...
+At first glance, extending `Nullable<T>` might be a good option, 
+but the `struct` constraint for `T` in it makes it impossible to use with the reference types.
+
+Using maybe monad also makes code more readable by reducing the nested code blocks, as you will see in the following examples.
+
+# How to use it
+
+Our hardworking teammates are trying to publish the Option library to [nuget](https://www.nuget.org/), but until then you can download/clone/fork this repository and build it locally. 
+
+> Option library targets **.NET Starndard 2.0** , so you can use it with **.NET Core 2.0+**, **.NET Framework 4.6.1+** and many other platforms. Please, see the [platform support.](https://github.com/dotnet/standard/blob/master/docs/versions/netstandard2.0.md)
```
Comment (rodoherty1) [inline README.md]: Typo in "Starndard"

File: README.md [hunk 6]
```diff
@@ -2,6 +2,118 @@
...
+
+# Examples
```
Comment (rodoherty1) [inline README.md]: Does C# do import statements?  If so, could you add the necessary imports so that the code below is compilable?
Do we have a ```samples``` folder?  If so, we should reference that here.

File: README.md [hunk 7]
```diff
@@ -2,6 +2,118 @@
...
+Maybe is a data type (also known as an option type) which is used to represent an optional value.
+
+### Maybe Monad
+Maybe monad is a generic type which contains either some value of it's underlying type or nothing.
+
+If you need more information you can check [wikibooks introduction](https://en.wikibooks.org/wiki/Haskell/Understanding_monads/Maybe) to maybe monad.
+
+# Why we need the Option library?
+
+> For simplicity you can think of `Nullable<T>` being closest thing to the maybe monad in C#. Although it doesn't represent all the characteristics of the maybe monad, it's sufficient enough to give the main idea.
```
Comment (rodoherty1) [inline README.md]: Is this paragraph a quotation?  Why is it greyed and indented?

File: README.md (hunk 1 above)
Comment (rodoherty1) [reply README.md]: Just been re-thinking this a little.  Perhaps these three sections could be removed and instead we would link to some excellent resources on the web whenever we you mention these terms elsewhere in the README?

**Review by rodoherty1 — CHANGES_REQUESTED** [review]
Great PR, @rustem-net !  A couple of comments which I'm happy to discuss in more detail any time.
Thanks again !!

File: README.md [hunk 8]
```diff
@@ -1,6 +1,142 @@
...
+For simplicity you can think of [`Nullable<T>`](https://msdn.microsoft.com/en-us/library/b3h38hb0(v=vs.110).aspx) being closest thing to the maybe monad in C#. Although it doesn't represent all the characteristics of the maybe monad, it's sufficient enough to give the main idea.
+
+At first glance, extending `Nullable<T>` might be a good option, 
+but the `struct` constraint for `T` in it makes it impossible to use with the reference types.
+
+# How to use it
+
+Our hardworking teammates are [trying to publish the Option Library](https://github.com/PaddyPowerBetfair/Option/issues/4) to [nuget](https://www.nuget.org/) , but until then you can download/clone/fork this repository and build it locally. 
+
+Option library targets **.NET Starndard 2.0** , so you can use it with **.NET Core 2.0+**, **.NET Framework 4.6.1+** and many other platforms. Please, see the [platform support.](https://github.com/dotnet/standard/blob/master/docs/versions/netstandard2.0.md)
```
Comment (rodoherty1) [inline README.md]: Typo in "Starndard"

File: README.md [hunk 9]
```diff
@@ -1,6 +1,142 @@
...
+var team = db.Teams.Find(t => t.Id == player.Id);
+if (team == null)
+    return;
+Console.WriteLine(team.Name);
+```
+
+### refactored using  Option Library, compact code which is easy to follow with safe chaining of steps 
+
+```cs
+Option
```
Comment (rodoherty1) [inline README.md]: This is much better.  But would it be possible to make it even clearer that the first two code snippets are _not_ a result from using our library.

Something like Before (in bold and red) and After (in bold and green).

Or say "Do this" followed by "Instead of this".

File: README.md [hunk 10]
```diff
@@ -1,6 +1,142 @@
...
+{
+    Console.WriteLine(player.Name);
+    var team = db.Teams.Find(t => t.Id == player.Id);
+    if (team != null)
+    {
+        Console.WriteLine(team.Name);
+    }
+}
+```
+### can be refactored using [Guard Clauses](https://refactoring.com/catalog/replaceNestedConditionalWithGuardClauses.html) , but still with null checking
```
Comment (rodoherty1) [inline README.md]: I would consider removing this.  The sample is similar to the first.

**Review by rodoherty1 — CHANGES_REQUESTED** [review]
Great job, Rustem!  I have requested a few more changes and then we can merge!

File: README.md [hunk 11]
```diff
@@ -1,8 +1,19 @@
 # Option
 
-Option Library is a C# representation of the [maybe monad](https://en.wikibooks.org/wiki/Haskell/Understanding_monads/Maybe) which provides a good way to deal with the exceptional cases by reducing the number of  [exception and null checkings](https://en.wikipedia.org/wiki/Exception_handling#Checked_exceptions) as well as the nested blocks in the code which leads to a more compact and more readable code.
+Option Library is a C# representation of the [maybe monad](https://en.wikibooks.org/wiki/Haskell/Understanding_monads/Maybe) which provides a good way to deal with the exceptional cases by reducing the number of  [exception and null checkings](https://en.wikipedia.org/wiki/Exception_handling#Checked_exceptions) as well as the nested blocks in the code which leads to a more compact and more readable code base.
```
Comment (sandorbari) [inline README.md]: link probably should say "exceptions and null checks"

File: README.md [hunk 12]
```diff
@@ -1,8 +1,19 @@
...
 
-### nested if statements : ugly code and hard to follow, known as [arrow anti pattern](http://wiki.c2.com/?ArrowAntiPattern)
+### **Use Option library** for *compact code* which is *easy to follow* with safe chaining of steps :
+
+```cs
+Option
+    .From(db.Players.FirstOrDefault(p => p.Name == keyword))
+    .Try(p => Console.WriteLine(p.Name))
+    .Select(p => db.Teams.Find(t => t.Id == p.TeamId))
+    .Try(t => Console.WriteLine(t.Name))
```
Comment (sandorbari) [inline README.md:12]: I like this example :)

File: README.md [hunk 13]
```diff
@@ -1,8 +1,19 @@
...
+```
+
+
+### **Instead of using nested if statements** which lead to *less readable code* and *hard to follow* [arrow anti pattern](http://wiki.c2.com/?ArrowAntiPattern)
```
Comment (sandorbari) [inline README.md]: should probably be "leads to"

**Review by rodoherty1 — APPROVED** [review]

**Comment by rodoherty1** [comment]
I think this is good to merge.  I think we can ignore the travis-ci error right now because #15 will hopefully fix this up (and if not we can raise a new issue and swarm on it).

Great stuff, @rustem-net.  Thanks again for your help on this!
